### PR TITLE
Implement transform_each

### PR DESCRIPTION
**Pull request by maikel**
This algorithm is a sequence sender adaptor. It takes as input parameters a sequence sender and a sender adaptor. 
The algorithm applies the sender adaptor to each emited item of the input sequence sender and returns the resulting sequence sender.

### CODE REVIEW

File: include/exec/sequence/transform_each.hpp [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+      _Receiver __receiver_;
+      _Adaptor __adaptor_;
+    };
+
+    template <class _ReceiverId, class _Adaptor>
+    struct __receiver {
+      using _Receiver = stdexec::__t<_ReceiverId>;
+
+      struct __t {
+        using is_receiver = void;
```
Comment (ericniebler) [inline include/exec/sequence/transform_each.hpp:36]:
```suggestion
        using is_receiver = void;
        using __id = __receiver;
```

File: include/exec/sequence/transform_each.hpp [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+          return stdexec::get_env(__self.__op_->__receiver_);
+        }
+      };
+    };
+
+    template <class _Sender, class _ReceiverId, class _Adaptor>
+    struct __operation {
+      using _Receiver = stdexec::__t<_ReceiverId>;
+
+      struct __t : __operation_base<_Receiver, _Adaptor> {
```
Comment (ericniebler) [inline include/exec/sequence/transform_each.hpp:79]:
```suggestion
      struct __t : __operation_base<_Receiver, _Adaptor> {
        using __id = __operation;
```

File: include/exec/sequence/transform_each.hpp [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+        template <class _Self, class _Env>
+        using __some_sender = __copy_cvref_t<
+          _Self,
+          __sequence_sndr::__unspecified_sender_of<
+            completion_signatures_of_t<__copy_cvref_t<_Self, _Sender>, _Env>>>;
+
+        template <class _Self, class _Env>
+        using completion_sigs_t = completion_signatures_of_t<
+          __call_result_t<_Adaptor&, __some_sender<_Self, _Sender>>,
```
Comment (ericniebler) [inline include/exec/sequence/transform_each.hpp:116]: Is this supposed to be:

```suggestion
          __call_result_t<_Adaptor&, __some_sender<_Self, _Env>>,
```

? Also, I don't think I really understand what's going on here. We are saying the completion signatures of `transform_each` are the completions of a sender type that is returned from the adaptor when passed ... some unspecified sender type? Doesn't it matter which sender is passed to the adaptor?

File: include/exec/sequence/transform_each.hpp [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+          -> completion_sigs_t<_Self, _Env>;
+
+        friend env_of_t<_Sender> tag_invoke(get_env_t, const __t& __self) noexcept {
+          return stdexec::get_env(__self.__sender_);
+        }
+      };
+    };
+
+    struct transform_each_t {
+      template <sender _Sender, class _Adaptor>
```
Comment (ericniebler) [inline include/exec/sequence/transform_each.hpp:143]: I think "`_Adaptor`" here is a sender adaptor closure, right? You can use `stdexec::__sender_adaptor_closure` to constrain the adaptor argument (or `stdexec::__sender_adaptor_closure_for` if you know the type of the sender that will be passed to the adaptor).

File: test/exec/sequence/test_transform_each.cpp [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+TEST_CASE(
+  "transform_each - transform sender applies adaptor",
+  "[sequence_senders][transform_each][empty_sequence]") {
+  int value = 0;
+  auto transformed = exec::transform_each(
+    stdexec::just(42), stdexec::then([&value](int x) noexcept { value = x; }));
+  auto op = exec::subscribe(transformed, next_rcvr{});
+  stdexec::start(op);
+  CHECK(value == 42);
+}
```
Comment (ericniebler) [inline test/exec/sequence/test_transform_each.cpp:48]: I'd love to see a test with dependently-typed senders.

**Review by ericniebler — CHANGES_REQUESTED** [review]

File: include/exec/sequence/transform_each.hpp (hunk 3 above)
Comment (maikel) [reply include/exec/sequence/transform_each.hpp:116]: We pass the items of the predecessor sequence to the adaptor. In the current design we chose to not list their concrete item type somewhere. I think it's inaccessible in this context here. One could imagine a design where a sequence does not specify completion signatures but it's item sender types instead.

You will find a similar use of an unspecified sender in the sequence sender concepts for this reason.

The choice to specify only completion signatures of items  has been made for convenience. But from this follows some consistency assumptions, i. e. If I pass an item with signatures S I always get an item with signatures R, indepently of the concrete item type.

A similar consistency assumption is present for set next.

File: include/exec/sequence/transform_each.hpp (hunk 4 above)
Comment (maikel) [reply include/exec/sequence/transform_each.hpp:143]: Yes I have missed this concrete vocabulary. Thank you

File: include/exec/sequence/transform_each.hpp (hunk 3 above)
Comment (ericniebler) [reply include/exec/sequence/transform_each.hpp:116]: > from this follows some consistency assumptions, i. e. If I pass an item with signatures S I always get an item with signatures R, indepently of the concrete item type.

This makes me a little uncomfortable. Algorithms are customizable, and I can imagine some implementations might change signatures slightly: an additional error, or a stop signal.

File: test/exec/sequence/test_transform_each.cpp [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  "[sequence_senders][transform_each]") {
+  int value = 0;
+  auto transformed = exec::transform_each(
+    stdexec::just(42), stdexec::then([&value](int x) noexcept { value = x; }));
+  auto op = exec::subscribe(transformed, next_rcvr{});
+  stdexec::start(op);
+  CHECK(value == 42);
+}
+
+#ifdef __cpp_lib_ranges
```
Comment (ericniebler) [inline test/exec/sequence/test_transform_each.cpp:65]: Could you rebase on `main` and use the new `STDEXEC_HAS_RANGES()` macro here? Thx.

File: include/exec/sequence/transform_each.hpp (hunk 3 above)
Comment (maikel) [reply include/exec/sequence/transform_each.hpp]: I have addressed that now with PR #1020 which is included here.

File: test/exec/sequence/test_transform_each.cpp (hunk 5 above)
Comment (maikel) [reply test/exec/sequence/test_transform_each.cpp]: I have included a test where I dependently-typed the sender which is sent by the sequence.

File: include/exec/sequence/transform_each.hpp [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+    template <class _Adaptor>
+    struct _NOT_CALLABLE_ADAPTOR_ { };
+
+    template <class _Item>
+    struct _WITH_ITEM_SENDER_ { };
+
+    template <class _Adaptor, class _Item>
+    auto __try_call(_Item*)
+      -> stdexec::__mexception<_NOT_CALLABLE_ADAPTOR_<_Adaptor&>, _WITH_ITEM_SENDER_<_Item>>;
```
Comment (ericniebler) [inline include/exec/sequence/transform_each.hpp:125]: i think we want to demangle the sender name here.

```suggestion
      -> stdexec::__mexception<_NOT_CALLABLE_ADAPTOR_<_Adaptor&>, _WITH_ITEM_SENDER_<stdexec::__name_of<_Item>>>;
```

**Review by ericniebler — APPROVED** [review]

File: include/exec/sequence/transform_each.hpp (hunk 7 above)
Comment (maikel) [reply include/exec/sequence/transform_each.hpp:125]: Good catch

File: test/exec/sequence/test_transform_each.cpp [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+#include "exec/sequence/transform_each.hpp"
+
+#include "exec/sequence/any_sequence_of.hpp"
+#include "exec/sequence/empty_sequence.hpp"
+#include "exec/sequence/iterate.hpp"
+#include "exec/sequence/ignore_all_values.hpp"
+#include <catch2/catch.hpp>
+
+#include <exec/on.hpp>
```
Comment (ericniebler) [inline test/exec/sequence/test_transform_each.cpp:26]:
```suggestion
#include <exec/env.hpp>
#include <exec/on.hpp>
```

**Pull request merged**